cmd/soroban-rpc/internal/integrationtest/infrastructure/client.go

```diff
@@ -0,0 +1,196 @@
+	var result methods.GetTransactionResponse
+	for i := 0; i < 60; i++ {
+		request := methods.GetTransactionRequest{Hash: hash}
+		err := client.CallResult(context.Background(), "getTransaction", request, &result)
+		assert.NoError(t, err)
+
+		if result.Status == methods.TransactionStatusNotFound {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		return result
+	}
+	t.Fatal("GetTransaction timed out")
+	return result
+}
+
+func SendSuccessfulTransaction(t *testing.T, client *Client, kp *keypair.Full, transaction *txnbuild.Transaction) methods.GetTransactionResponse {
+	tx, err := transaction.Sign(StandaloneNetworkPassphrase, kp)
+	assert.NoError(t, err)
+	b64, err := tx.Base64()
+	assert.NoError(t, err)
+
+	request := methods.SendTransactionRequest{Transaction: b64}
+	var result methods.SendTransactionResponse
+	assert.NoError(t, client.CallResult(context.Background(), "sendTransaction", request, &result))
+
+	expectedHashHex, err := tx.HashHex(StandaloneNetworkPassphrase)
+	assert.NoError(t, err)
+
+	assert.Equal(t, expectedHashHex, result.Hash)
+	if !assert.Equal(t, stellarcore.TXStatusPending, result.Status) {
+		var txResult xdr.TransactionResult
+		err := xdr.SafeUnmarshalBase64(result.ErrorResultXDR, &txResult)
+		assert.NoError(t, err)
+		t.Logf("error: %#v\n", txResult)
+	}
+	assert.NotZero(t, result.LatestLedger)
+	assert.NotZero(t, result.LatestLedgerCloseTime)
+
+	response := getTransaction(t, client, expectedHashHex)
+	if !assert.Equal(t, methods.TransactionStatusSuccess, response.Status) {
+		var txResult xdr.TransactionResult
+		assert.NoError(t, xdr.SafeUnmarshalBase64(response.ResultXdr, &txResult))
+		t.Logf("error: %#v\n", txResult)
+
+		var txMeta xdr.TransactionMeta
+		assert.NoError(t, xdr.SafeUnmarshalBase64(response.ResultMetaXdr, &txMeta))
+
+		if txMeta.V == 3 && txMeta.V3.SorobanMeta != nil {
+			if len(txMeta.V3.SorobanMeta.Events) > 0 {
+				t.Log("Contract events:")
+				for i, e := range txMeta.V3.SorobanMeta.Events {
+					t.Logf("  %d: %s\n", i, e)
+				}
+			}
+
+			if len(txMeta.V3.SorobanMeta.DiagnosticEvents) > 0 {
+				t.Log("Diagnostic events:")
+				for i, d := range txMeta.V3.SorobanMeta.DiagnosticEvents {
+					t.Logf("  %d: %s\n", i, d)
+				}
+			}
+		}
+	}
+
+	require.NotNil(t, response.ResultXdr)
+	assert.Greater(t, response.Ledger, result.LatestLedger)
+	assert.Greater(t, response.LedgerCloseTime, result.LatestLedgerCloseTime)
+	assert.GreaterOrEqual(t, response.LatestLedger, response.Ledger)
+	assert.GreaterOrEqual(t, response.LatestLedgerCloseTime, response.LedgerCloseTime)
+	return response
+}
+
+func SimulateTransactionFromTxParams(t *testing.T, client *Client, params txnbuild.TransactionParams) methods.SimulateTransactionResponse {
+	savedAutoIncrement := params.IncrementSequenceNum
+	params.IncrementSequenceNum = false
+	tx, err := txnbuild.NewTransaction(params)
+	require.NoError(t, err)
+	params.IncrementSequenceNum = savedAutoIncrement
```

Comment:
Isn't this passed by value and so it doesn't need this preservation behavior?

Reply:
It is an interface which is passed by reference

Reply:
This code predates the refactoring though (although it's understandably hard to see with so many changes)

Reply:
Yeah it's hard to tell what moved and what changed but thanks for clarifying!